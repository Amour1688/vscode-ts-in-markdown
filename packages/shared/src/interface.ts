export interface ServerInitializationOptions {
  appRoot: string;
}
