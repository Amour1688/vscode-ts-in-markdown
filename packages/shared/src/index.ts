import * as path from 'path';
import { URI } from 'vscode-uri';
import { Language } from './interface';

export * from './interface';
export * from './parse';

export function loadTypeScript(
  appRoot: string,
): typeof import('typescript/lib/tsserverlibrary') {
  const tsPath = path.join(appRoot, 'extensions', 'node_modules', 'typescript');
  // eslint-disable-next-line
  return require(tsPath);
}

export function uriToFsPath(uri: string) {
  return URI.parse(uri).fsPath;
}

export function fsPathToUri(fsPath: string) {
  return URI.file(fsPath).toString();
}

export function normalizeFileName(fileName: string) {
  return uriToFsPath(fsPathToUri(fileName));
}

export function toVirtualPath(fileName: string, index: number = 0, lang?: Language) {
  return `${fileName}.__${index}.${lang || 'tsx'}`;
}

export function filterEmpty<T>(value: T | null | undefined): value is T {
  return value !== null && value !== undefined;
}

export const languageIdMap = {
  ts: 'typescript',
  js: 'javascript',
  tsx: 'typescriptreact',
  jsx: 'javascriptreact',
};
