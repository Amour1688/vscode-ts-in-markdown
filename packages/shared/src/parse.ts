/* eslint-disable no-plusplus */
export interface Position {
  offset: number;
  line: number;
  column: number;
}

export interface Location {
  start?: Position;
  end?: Position;
}

export interface SourceLocation {
  locations: Location[];
  source?: string;
}

export const locationMap = new Map<string, Location[]>(); // virtual path

function isCode(source: string, lang: string, start: number) {
  return source.slice(start, start + lang.length) === lang;
}

function isBlock(source: string, cursor: number) {
  return source.slice(cursor, cursor + 3) === '```';
}

export function parse(_source: string) {
  let i = 0;
  let line = 0;
  let column = 0;
  let shouldReplace = true;
  let location: Location | undefined = {};
  const sourceLocation: SourceLocation = {
    locations: [],
  };
  const results: string[] = [];

  while (i < _source.length) {
    const isLineBreak = _source[i] === '\n';
    results[i] = shouldReplace && !isLineBreak ? ' ' : _source[i];

    i++;
    column++;
    if (isLineBreak) {
      line++;
      column = 0;
    }

    if (isBlock(_source, i) && (isCode(_source, 'jsx', i + 3) || isCode(_source, 'tsx', i + 3))) {
      // ```tsx live=true
      while (i < _source.length && _source[i] !== '\n') {
        results[i] = ' ';
        i++;
      }
      shouldReplace = false;
      location = {
        start: {
          line: line + 1,
          column,
          offset: i + 1,
        },
      };
    } else if (location?.start?.line && isBlock(_source, i + 3)) {
      while (i < _source.length && _source[i] !== '\n') {
        results[i] = _source[i];
        i++;
      }
      location.end = {
        line,
        column,
        offset: i,
      };
      shouldReplace = true;
      sourceLocation.locations.push(location);
      location = undefined;
    }
    sourceLocation.source = results.join('');
  }
  return sourceLocation;
}

export function parseMarkdown(fileName: string, content: string) {
  const { source = '', locations } = parse(content);
  locationMap.set(fileName, locations);
  return source;
}
