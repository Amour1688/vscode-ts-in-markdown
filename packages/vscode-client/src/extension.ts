import * as vscode from 'vscode';
import * as lsp from 'vscode-languageclient/node';
import * as path from 'path';
import type { ServerInitializationOptions } from '@dali/shared';

let client: lsp.LanguageClient;

export async function activate(context: vscode.ExtensionContext) {
  const serverModule = context.asAbsolutePath(path.join('node_modules', '@dali', 'vscode-server', 'out', 'server.js'));
  const debugOptions = { execArgv: ['--nolazy', '--inspect=6009'] };

  const serverOptions: lsp.ServerOptions = {
    run: { module: serverModule, transport: lsp.TransportKind.ipc },
    debug: {
      module: serverModule,
      transport: lsp.TransportKind.ipc,
      options: debugOptions,
    },
  };

  const serverInitOptions: ServerInitializationOptions = {
    appRoot: vscode.env.appRoot,
  };

  const clientOptions: lsp.LanguageClientOptions = {
    documentSelector: [{ scheme: 'file', language: 'markdown' }],
    initializationOptions: serverInitOptions,
  };

  client = new lsp.LanguageClient(
    'Dali Markdown',
    'markdown',
    serverOptions,
    clientOptions,
  );
  context.subscriptions.push(client.start());

  return client;
}

export function deactivate(): Thenable<void> | undefined {
  return client?.stop();
}
