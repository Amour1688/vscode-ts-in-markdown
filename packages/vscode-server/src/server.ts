import {
  createConnection,
  ProposedFeatures,
  TextDocumentSyncKind,
  TextDocuments,
  DidChangeConfigurationNotification,
} from 'vscode-languageserver/node';
import { TextDocument } from 'vscode-languageserver-textdocument';
import type { ServerInitializationOptions } from '@ts-in-markdown/shared';
import { loadTypeScript, uriToFsPath } from '@ts-in-markdown/shared';
import type {
  InitializeResult,
  InitializeParams,
} from 'vscode-languageserver/node';
import { register as registerApiFeatures } from './registers/registerApiFeatures';
import { register as registerMdTsFeatures } from './features/mdTsFeatures';

let options: ServerInitializationOptions;
let folders: string[] = [];

const connection = createConnection(ProposedFeatures.all);
const documents = new TextDocuments(TextDocument);

connection.onInitialize(onInitialize);
connection.onInitialized(onInitialized);
connection.listen();
documents.listen(connection);

function onInitialize(params: InitializeParams) {
  options = params.initializationOptions;
  folders = params.workspaceFolders
    ? params.workspaceFolders
      .map((folder) => folder.uri)
      .filter((uri) => uri.startsWith('file:/'))
      .map((uri) => uriToFsPath(uri))
    : [];
  const result: InitializeResult = {
    capabilities: {
      textDocumentSync: TextDocumentSyncKind.Incremental,
    },
  };

  return result;
}

async function onInitialized() {
  registerApiFeatures(connection);

  registerMdTsFeatures(
    connection,
    loadTypeScript(options.appRoot),
    documents,
    folders,
  );
  connection.client.register(
    DidChangeConfigurationNotification.type,
    undefined,
  );
}
