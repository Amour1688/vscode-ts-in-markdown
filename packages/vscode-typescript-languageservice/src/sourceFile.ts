import { parse, SourceLocation } from '@dali/shared';

const location = new Map<string, SourceLocation>(); // virtual path

export function createSourceFile(fileName: string, content: string) {
  const { source = '', ...rest } = parse(content);
  location.set(fileName, rest);
  return source;
}

export { location };
