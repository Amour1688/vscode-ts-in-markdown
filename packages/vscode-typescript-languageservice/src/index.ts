import { LanguageServiceHost } from 'typescript';
import { TextDocument } from 'vscode-languageserver-textdocument';
import * as path from 'path';
import * as fg from 'fast-glob';
import {
  fsPathToUri,
  uriToFsPath,
  normalizeFileName,
  toVirtualPath,
  toRealFilePath,
} from '@ts-in-markdown/shared';
import { TextDocuments } from 'vscode-languageserver/node';
import * as hover from './languageFeatures/hover';
import * as definitions from './languageFeatures/definitions';
import * as completions from './languageFeatures/completions';
import * as completionResolve from './languageFeatures/completionResolve';
import * as typeDefinition from './languageFeatures/typeDefinition';
import * as formatting from './languageFeatures/formatting';
import * as folding from './languageFeatures/folding';
import * as references from './languageFeatures/references';
import { parseSourceFile } from './sourceFile';

export * from './sourceFile';

export function createLanguageService(
  ts: typeof import('typescript/lib/tsserverlibrary'),
  documents: TextDocuments<TextDocument>,
  folders: string[],
) {
  let projectVersion = 0;
  const tsConfigNames = ['tsconfig.json'];
  const snapshots = new Map<string, {
    version: string
    snapshot: ts.IScriptSnapshot
  }>();
  const tsConfigSet = new Set(
    folders
      .map((folder) => ts.sys.readDirectory(folder, tsConfigNames, undefined, ['**/*']))
      .flat(),
  );
  const tsConfigs = [...tsConfigSet].filter((tsConfig) => tsConfigNames.includes(path.basename(tsConfig)));
  let parsedCommandLine: ts.ParsedCommandLine;
  const mdMap = new Map<string, { version: number, fileName: string }>();
  const virtualMap = new Map<string, string>();
  const tsFiles = new Map<string, { version: number; fileName: string }>();
  const documentsMap = new Map<string, { version: number, document: TextDocument }>();

  update();

  const host = createTsLanguageServiceHost();
  const languageService = ts.createLanguageService(
    host,
    ts.createDocumentRegistry(),
  );

  return {
    dispose,
    doHover: hover.register(languageService, getTextDocument, ts),
    doCompletion: completions.register(languageService, getTextDocument),
    doCompletionResolve: completionResolve.register(languageService, getTextDocument),
    doFormatting: formatting.register(languageService, getTextDocument),
    doFolding: folding.register(languageService, getTextDocument),
    fineTypeDefinition: typeDefinition.register(languageService, getTextDocument),
    findDefinitions: definitions.register(languageService, getTextDocument),
    findReferences: references.register(languageService, getTextDocument),
    onDocumentUpdate,
    update,
  };

  function update() {
    const mds = folders
      .map((folder) => [...fg.sync(`${folder}/components/**/*.md`), ...fg.sync(`${folder}/src/**/*.md`)])
      .flat();

    const mdSet = new Set(mds);
    for (const [markdown] of mdMap) {
      if (!mdSet.has(markdown)) {
        mdMap.delete(markdown);
      }
    }

    let change = false;
    mds.forEach((markdown) => {
      const virtualName = toVirtualPath(markdown);
      if (!mdMap.has(virtualName)) {
        mdMap.set(markdown, { version: 0, fileName: virtualName });
        virtualMap.set(virtualName, markdown);
        change = true;
      }
    });

    parsedCommandLine = createParsedCommandLine(
      ts,
      tsConfigs[0],
    );

    const fileNames = new Set(parsedCommandLine.fileNames);
    for (const [fileName] of tsFiles) {
      if (!fileNames.has(fileName)) {
        tsFiles.delete(fileName);
      }
    }

    for (const fileName of parsedCommandLine.fileNames) {
      if (!tsFiles.has(fileName)) {
        tsFiles.set(fileName, {
          version: 0,
          fileName,
        });
        change = true;
      }
    }

    if (change) {
      projectVersion++;
    }
  }

  function dispose() {
    languageService.dispose();
  }

  function createTsLanguageServiceHost() {
    const serviceHost: LanguageServiceHost = {
      // ts
      getNewLine: () => ts.sys.newLine,
      useCaseSensitiveFileNames: () => ts.sys.useCaseSensitiveFileNames,
      readFile: ts.sys.readFile,
      writeFile: ts.sys.writeFile,
      directoryExists: ts.sys.directoryExists,
      getDirectories: ts.sys.getDirectories,
      readDirectory: ts.sys.readDirectory,
      realpath: ts.sys.realpath,
      fileExists,
      getCompilationSettings: () => parsedCommandLine.options,
      getProjectVersion: () => `${projectVersion}`,
      getScriptFileNames: () => [
        ...parsedCommandLine.fileNames,
        ...([...mdMap.values()].map(({ fileName }) => fileName)),
      ],
      getScriptVersion,
      getCurrentDirectory: () => path.dirname(tsConfigs[0]),
      getScriptSnapshot,
      getDefaultLibFileName: (options) => ts.getDefaultLibFilePath(options),
    };
    return serviceHost;
  }

  function fileExists(_fileName: string) {
    const fileName = normalizeFileName(
      ts.sys.realpath?.(_fileName) ?? _fileName,
    );
    return !!ts.sys.fileExists?.(fileName);
  }

  function getScriptVersion(fileName: string) {
    const virtual = virtualMap.get(fileName);
    return `${virtual ? mdMap.get(virtual)?.version : tsFiles.get(fileName)?.version || 0}`;
  }

  function getScriptSnapshot(fileName: string) {
    const version = getScriptVersion(fileName);
    const cache = snapshots.get(fileName);
    if (cache?.version === version) {
      return cache.snapshot;
    }
    const text = getScriptText(fileName);
    if (text !== undefined) {
      const snapshot = ts.ScriptSnapshot.fromString(text);
      snapshots.set(fileName, {
        version: `${version}`,
        snapshot,
      });
      return ts.ScriptSnapshot.fromString(text);
    }
  }

  function getScriptText(fileName: string) {
    const doc = documents.get(fsPathToUri(toRealFilePath(fileName)));
    if (doc) {
      return virtualMap.has(fileName) ? parseSourceFile(fileName, doc.getText()) : doc.getText();
    }
    if (ts.sys.fileExists(fileName)) {
      return ts.sys.readFile(fileName, 'utf8');
    }
  }

  function getTextDocument(uri: string) {
    const fileName = uriToFsPath(uri);
    if (!languageService.getProgram()?.getSourceFile(fileName)) {
      return;
    }
    const version = host.getScriptVersion(fileName);
    const prev = documentsMap.get(fileName);
    if (prev?.version !== Number(version)) {
      const scriptSnapshot = host.getScriptSnapshot(fileName);
      if (scriptSnapshot) {
        const scriptText = scriptSnapshot.getText(0, scriptSnapshot.getLength());
        const newVersion = prev?.version ? prev.version + 1 : 0;
        const document = TextDocument.create(uri, 'typescript', newVersion, scriptText);
        documentsMap.set(fileName, {
          version: newVersion,
          document,
        });
      }
    }
    return documentsMap.get(fileName)?.document;
  }

  function onDocumentUpdate(document: TextDocument) {
    const fsPath = uriToFsPath(document.uri);
    const markdown = mdMap.get(fsPath);
    const fileName = markdown ? markdown.fileName : fsPath;
    const snapshot = snapshots.get(fileName);
    if (snapshot) {
      const snapshotLength = snapshot.snapshot.getLength();
      const documentText = markdown ? parseSourceFile(fileName, document.getText()) : document.getText();
      if (
        snapshotLength === documentText.length
        && snapshot.snapshot.getText(0, snapshotLength) === documentText
      ) {
        return;
      }
    }
    const file = markdown ?? tsFiles.get(fsPath);
    if (file) {
      file.version++;
      projectVersion++;
    }
  }
}

function createParsedCommandLine(
  ts: typeof import('typescript/lib/tsserverlibrary'),
  tsConfig: string,
) {
  const parseConfigHost: ts.ParseConfigHost = {
    ...ts.sys,
    readDirectory: ts.sys.readDirectory,
  };
  const realTsConfig = ts.sys.realpath!(tsConfig);
  const config = ts.readJsonConfigFile(realTsConfig, ts.sys.readFile);
  const content = ts.parseJsonSourceFileConfigFileContent(
    config,
    parseConfigHost,
    path.dirname(realTsConfig),
    {},
    path.basename(realTsConfig),
  );
  content.options.outDir = undefined;
  content.fileNames = content.fileNames.map(normalizeFileName);
  return content;
}
