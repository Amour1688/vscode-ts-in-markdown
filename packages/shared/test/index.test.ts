import * as fs from 'fs'
import * as path from 'path'
import { parse } from '../src/parse';

const a = fs.readFileSync(path.join(__dirname, './fixtures/a.md')).toString()

describe('parseMarkdown', () => {
  it('should parse markdown', () => {
    const result = parse(a);
    expect(result).toMatchSnapshot();
  });
});
