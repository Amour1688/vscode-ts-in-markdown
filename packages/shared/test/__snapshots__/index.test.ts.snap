// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseMarkdown should parse markdown 1`] = `
Array [
  Object {
    "content": "    

      
console.log(1)",
    "language": "jsx",
    "location": Object {
      "end": Object {
        "column": 12,
        "line": 3,
        "offset": 27,
      },
      "start": Object {
        "column": 0,
        "line": 3,
        "offset": 13,
      },
    },
  },
  Object {
    "content": "    

      
              
   

     

      
const a = 1",
    "language": "tsx",
    "location": Object {
      "end": Object {
        "column": 9,
        "line": 9,
        "offset": 58,
      },
      "start": Object {
        "column": 0,
        "line": 9,
        "offset": 47,
      },
    },
  },
  Object {
    "content": "    

      
              
   

     

      
           
   

      

      
const b = 2",
    "language": "tsx",
    "location": Object {
      "end": Object {
        "column": 9,
        "line": 15,
        "offset": 90,
      },
      "start": Object {
        "column": 0,
        "line": 15,
        "offset": 79,
      },
    },
  },
  Object {
    "content": "    

      
              
   

     

      
           
   

      

      
           
   

      
     
const a = 1",
    "language": "js",
    "location": Object {
      "end": Object {
        "column": 9,
        "line": 20,
        "offset": 120,
      },
      "start": Object {
        "column": 0,
        "line": 20,
        "offset": 109,
      },
    },
  },
  Object {
    "content": "    

      
              
   

     

      
           
   

      

      
           
   

      
     
           
   

      
     
const a = 1",
    "language": "ts",
    "location": Object {
      "end": Object {
        "column": 9,
        "line": 25,
        "offset": 150,
      },
      "start": Object {
        "column": 0,
        "line": 25,
        "offset": 139,
      },
    },
  },
]
`;
